transports/quic/src/endpoint.rs

```diff
@@ -0,0 +1,459 @@
+use crate::crypto::{Crypto, CryptoConfig};
+use crate::muxer::QuicMuxer;
+use crate::{QuicConfig, QuicError};
+use fnv::FnvHashMap;
+use futures::channel::{mpsc, oneshot};
+use futures::prelude::*;
+use quinn_proto::crypto::Session;
+use quinn_proto::generic::{ClientConfig, ServerConfig};
+use quinn_proto::{
+    ConnectionEvent, ConnectionHandle, DatagramEvent, EcnCodepoint, EndpointEvent, Transmit,
+};
+use std::collections::VecDeque;
+use std::io::IoSliceMut;
+use std::mem::MaybeUninit;
+use std::net::SocketAddr;
+use std::pin::Pin;
+use std::sync::Arc;
+use std::task::{Context, Poll};
+use std::time::Instant;
+use udp_socket::{RecvMeta, SocketType, UdpCapabilities, UdpSocket, BATCH_SIZE};
+
+/// Message sent to the endpoint background task.
+#[derive(Debug)]
+enum ToEndpoint<C: Crypto> {
+    /// Instructs the endpoint to start connecting to the given address.
+    Dial {
+        /// UDP address to connect to.
+        addr: SocketAddr,
+        /// The remotes public key.
+        public_key: C::PublicKey,
+        /// Channel to return the result of the dialing to.
+        tx: oneshot::Sender<Result<QuicMuxer<C>, QuicError>>,
+    },
+    /// Sent by a `quinn_proto` connection when the endpoint needs to process an event generated
+    /// by a connection. The event itself is opaque to us.
+    ConnectionEvent {
+        connection_id: ConnectionHandle,
+        event: EndpointEvent,
+    },
+    /// Instruct the endpoint to transmit a packet on its UDP socket.
+    Transmit(Transmit),
+}
+
+#[derive(Debug)]
+pub struct TransportChannel<C: Crypto> {
+    tx: mpsc::UnboundedSender<ToEndpoint<C>>,
```

Comment:
Is there a proof that this always remains bounded?

Reply:
same here, this is the channel from libp2p not from the socket. other than erroring if the queue is full and trying to dial I don't see what we can do

Reply:
That makes sense.  FYI I believe much of the complexity in my original PR was from avoiding unbounded channels at all costs, which required me to use locking instead of message-passing.

Reply:
Marking as resolved

Reply:
> same here, this is the channel from libp2p not from the socket. other than erroring if the queue is full and trying to dial I don't see what we can do

I don't understand why the direction (libp2p -> socket) matters in regards to bounded or unbounded. Can you expand on that @dvc94ch?

Reply:
you want a bounded channel to prevent dos attacks. you can't prevent a user from dos'ing himself, he can just exit the application.